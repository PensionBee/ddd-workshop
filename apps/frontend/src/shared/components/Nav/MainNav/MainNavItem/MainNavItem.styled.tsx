import styled, { css } from "styled-components";
import { scaleIn, scaleOut } from "../../Nav.styled";
import { LargeSystemIcon } from "../../../Svg/Svg";
import { theme } from "@shared/styling/theme";

type StyledItemProps = {
  $isTabMenu?: boolean;
};

const tabLinkStyles = css`
  display: flex;
  margin-bottom: 20px;
  color: ${() => theme.colors.grey1};
  &:hover,
  &:focus-visible {
    color: ${() => theme.colors.black};
  }
  &:focus-visible {
    outline: 3px solid ${() => theme.colors.blue100};
  }
`;

type StyledTriggerProps = React.HTMLProps<HTMLButtonElement> & StyledItemProps;

export const StyledTrigger = styled.button<StyledTriggerProps>`
  ${({ $isTabMenu }) => $isTabMenu && tabLinkStyles}
  cursor: pointer;
  user-select: none;
  transition: color 200ms ease;
  display: flex;
  align-items: center;
  justify-content: space-between;
  gap: 10px;
  color: ${() => theme.colors.black};
  &:hover,
  &:focus-visible {
    color: ${() => theme.colors.grey1};
  }
  &:focus-visible {
    outline: 3px solid ${() => theme.colors.blue100};
    border-radius: 4px;
  }
`;

export const StyledCaret = styled(LargeSystemIcon)`
  position: relative;
  top: 1;
  transform: rotate(-180deg);
  transition: transform 250ms ease;
  [data-state="open"] & {
    transform: rotate(0deg);
  }
`;

export type StyledLinkProps = StyledItemProps & {
  $isHighlighted?: boolean;
};

export const StyledLink = styled.a<StyledLinkProps>`
  ${({ $isTabMenu }) => $isTabMenu && tabLinkStyles}
  display: flex;
  transition: color 200ms ease;
  color: ${() => theme.colors.black};
  &:hover,
  &:focus-visible {
    color: ${() => theme.colors.grey1};
  }
  &:focus-visible {
    outline: 3px solid ${() => theme.colors.blue100};
    border-radius: 4px;
  }
  ${({ $isHighlighted }) =>
    $isHighlighted &&
    css`
      color: ${() => theme.colors.black};
      &::after {
        content: "";
        border-bottom: 1px solid ${theme.colors.yellow};
        width: 100%;
        position: absolute;
        bottom: 0;
        margin-bottom: -1px;
      }
    `}
`;

export const StyledContent = styled.div`
  position: absolute;
  top: 26px;
  left: -20px;
  animation-duration: 250ms;
  animation-timing-function: ease;
  transform-origin: top center;
  margin-top: 8px;
  width: 100%;
  background-color: white;
  overflow: hidden;
  box-shadow: 0px 5px 15px 0px rgb(0 0 0 / 18%);
  border-radius: 8px;
  transition: width, height, 300ms ease;
  @media only screen and (min-width: 600px) {
    width: auto;
  }
  &[data-state="open"] {
    animation: ${scaleIn} 200ms ease;
  }
  &[data-state="closed"] {
    animation: ${scaleOut} 200ms ease;
  }
`;

export const StyledItem = styled.li`
  position: relative;
`;
