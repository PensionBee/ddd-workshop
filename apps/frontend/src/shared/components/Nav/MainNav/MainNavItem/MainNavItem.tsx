import type { NavMenuItem } from "../../Nav.types";
import { DropdownMenu } from "../../DropdownMenu/DropdownMenu";
import {
  StyledCaret,
  StyledContent,
  StyledItem,
  StyledTrigger,
} from "./MainNavItem.styled";
import { useRef } from "react";

export type MainNavItemProps = NavMenuItem & {
  $isTabMenu?: boolean;
  $isOpen?: boolean;
  toggleMenu?: (key: NavMenuItem["children"]) => void;
  handleCloseMenu?: () => void;
};

export const MainNavItem: React.FC<MainNavItemProps> = ({
  children,
  subList,
  $isTabMenu,
  $isOpen,
  toggleMenu,
  handleCloseMenu,
  ...props
}) => {
  const triggerRef = useRef<HTMLButtonElement | null>(null);
  const contentRef = useRef<HTMLDivElement | null>(null);

  const handleTriggerClick = () => {
    toggleMenu?.(children);
  };

  const handleCloseDropdown = () => {
    handleCloseMenu?.();
    triggerRef.current?.focus();
  };

  const handleArrowDownOpenMenu = (
    event: React.KeyboardEvent<HTMLButtonElement>
  ) => {
    if (event.key === "ArrowDown") {
      event.preventDefault();
      handleTriggerClick();
    }
  };

  return (
    <StyledItem
      data-rc="MainNavItem"
      data-state={$isOpen ? "open" : "closed"}
      {...props}
    >
      <StyledTrigger
        ref={triggerRef}
        $isTabMenu={$isTabMenu}
        onClick={handleTriggerClick}
        onKeyUp={handleArrowDownOpenMenu}
      >
        {children}
        <StyledCaret icon="ChevronYellow" aria-hidden />
      </StyledTrigger>
      <StyledContent ref={contentRef}>
        {$isOpen && subList && (
          <DropdownMenu
            subList={subList}
            handleCloseDropdown={handleCloseDropdown}
          />
        )}
      </StyledContent>
    </StyledItem>
  );
};
