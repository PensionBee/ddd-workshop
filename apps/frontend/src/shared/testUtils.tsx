import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import type { Breakpoints } from "./styling/breakpoints";
import { BREAKPOINTS } from "./styling/breakpoints";
import { theme } from "./styling/theme";
import { vi } from "vitest"

type ReactTestingLibraryProviderProps = {
  children: React.ReactNode;
};

// eslint-disable-next-line react-refresh/only-export-components
const ReactTestingLibraryProviders: React.FC<
  ReactTestingLibraryProviderProps
> = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const customRender = (ui: React.ReactElement, options?: any) => {
  return render(ui, { wrapper: ReactTestingLibraryProviders, ...options });
};

// re-export everything
// eslint-disable-next-line react-refresh/only-export-components
export * from "@testing-library/react";

// override render method
export { customRender as render };

export const sleep = (delay: number) => {
  return new Promise((resolve) => setTimeout(resolve, delay));
};

export const mockMediaMatches = (breakpoint: Breakpoints) => {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: (query: string) => {
      const breakpointValue = BREAKPOINTS[breakpoint];
      const queryValue = parseInt(query.replace(/^\D+/g, ""));
      const breakpointValues = [...Object.values(BREAKPOINTS)];
      if (!breakpointValues.includes(queryValue)) {
        throw new Error(
          `Invalid breakpoint value: ${queryValue}, this likely means there has been a change to the way that breakpoints are setup.`
        );
      }
      return {
        matches: queryValue <= breakpointValue,
        addEventListener: vi.fn(),
        removeEventListener: vi.fn(),
      };
    },
  });
};

export const routerRender = (ui: React.ReactNode, pathname = "/") => {
  return customRender(
    <MemoryRouter initialEntries={[{ pathname }]}>{ui}</MemoryRouter>
  );
};
