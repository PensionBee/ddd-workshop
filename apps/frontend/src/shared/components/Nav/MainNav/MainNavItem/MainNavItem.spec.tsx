import { routerRender } from "@shared/testUtils";
import { MainNavItem } from "./MainNavItem";

const ITEM_CONTENT = "Item Content";

describe("MainNavItem", () => {
  it("should render correctly", () => {
    const { getByText, asFragment } = routerRender(
      <MainNavItem>{ITEM_CONTENT}</MainNavItem>
    );

    expect(getByText(ITEM_CONTENT)).toBeVisible();
    expect(asFragment()).toMatchSnapshot();
  });
});
