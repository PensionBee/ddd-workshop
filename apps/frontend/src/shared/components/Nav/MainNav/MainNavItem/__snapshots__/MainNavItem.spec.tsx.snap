// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainNavItem should render correctly 1`] = `
<DocumentFragment>
  <li
    class="sc-csuSiG hrlAfi"
    data-rc="MainNavItem"
    data-state="closed"
  >
    <button
      class="sc-pyfCe hMxGWU"
    >
      Item Content
      <span
        aria-hidden="true"
        class="sc-gswNZR iagQga sc-jrcTuL hgtJKb"
        data-rc="LargeSystemIcon.ChevronYellow"
      >
        <svg>
          chevron-yellow.svg
        </svg>
      </span>
    </button>
    <div
      class="sc-ipEyDJ cPLSnD"
    />
  </li>
</DocumentFragment>
`;
